R/mlr_graphs_elements.R

```diff
@@ -0,0 +1,163 @@
+#' * Imputes `numeric` features using [`PipeOpImputeHist`] and [`PipeOpMissInd`]
+#' * Imputes `factor` features using [`PipeOpImputeNewlvl`]
+#' * Encodes `factors` using `one-hot-encoding`. Factors with a cardinality > max_cardinality` are
+#'   collapsed using [`PipeOpCollapseFactors`].
+#' * If `scaling`, numeric features are scaled to mean 0 and standard deviation 1.
+#'
+#' The graph is built conservatively, i.e. the function always tries to assure everything works.
+#' If a learner is provided, some steps can be left out, i.e. if the learner can deal with
+#' factor variables, no encoding is performed.
+#'
+#' @param task [`Task`] \cr
+#'   A task to create a robustifying pipeline for.
+#'   Optional, if omitted, the full pipeline is created.
+#' @param learner [`Learner`] \cr
+#'   A learner to create a robustifying pipeline for. Optional, if omitted,
+#'   a more conservative pipeline is built.
+#' @param impute_missings [`logical`] | [`NULL`] \cr
+#'   Should missing values be imputed? Defaults to `NULL`, i.e imputes if the task has
+#'   missing values and the learner can not handle them.
+#' @param factors_to_numeric [`logical`] | [`NULL`] \cr
+#'   Should factors be encoded? Defaults to `NULL`, i.e encodes if the task has factors
+#'   and the learner can not handle factors.
+#' @param scaling [`logical`] \cr
+#'   Should the data be scaled? Defaults to true.
+#' @param max_cardinality [`integer`] \cr
+#'   Maximum number of factor levels allowed. See above. Default: 1000.
+#' @return [`Graph`]
+#' @export
+#' @examples
+#' library(mlr3)
+#' lrn = lrn("regr.rpart")
+#' task = mlr_tasks$get("boston_housing")
+#' gr = robustify_pipeline(task, lrn) %>>% po("learner", lrn)
+#' resample(task,GraphLearner$new(gr), rsmp("holdout"))
+robustify_pipeline = function(task = NULL, learner = NULL, impute_missings = NULL,
+  factors_to_numeric = NULL, scaling = TRUE, max_cardinality = 1000) {
+
+  cols_by_type = function(types) {
+    task$feature_types[get("type") %in% types,]
+  }
+  if (!is.null(task)) assert_task(task)
+  if (!is.null(learner)) assert_learner(learner)
+  if (is.null(impute_missings)) impute_missings = (is.null(task) || any(task$missings())) && (is.null(learner) || !"missings" %in% learner$properties)
+  else assert_flag(impute_missings)
+  if (is.null(factors_to_numeric)) factors_to_numeric = (is.null(task) || nrow(cols_by_type("factor")) > 0) && (is.null(learner) || !"factors" %in% learner$properties)
+  else assert_flag(impute_missings)
+  assert_flag(scaling)
+  assert_count(max_cardinality)
+
+  # FIXME
+  # This should additionally do the following thins:
+  # - Convert posixct to numeric using new pipeop
+  # - Auto-convert character to factor? Perhaps improve later with text po's
+  # - Logical to integer conversion
+
+  # If given a task, only treat actually existing column types
+  if (!is.null(task)) {
+    pos = list()
+    if ("factor" %in% learner$feature_types && nrow(cols_by_type("factor")) > 0)
```

Comment:
what is our behaviour if learner can not handle factors? we just ignore new factors because they get converted to numerics? In that case we should go by `factors_to_numeric` here. But maybe we want to fix unseen factors to NAs and then impute in any case? Then just go by `is.null(task) || nrow(cols_by_type("factor"))`